## Part 2 - Computing the position of a particle within an electrostatic precipitator

An electrostatic precipitator is a device that is typically used to capture and remove harmful particles from the exhaust of industrial and/or chemical processes. The device works in two stages:
1. Particles in the exhaust are charged (either positively or negatively) 
2. The particles pass through an air channel where high voltage electrodes attract the charged particles (see figure 1 below). The electrodes create an electric field that exerts a coulomb force on charged particles. The particles move towards the electrodes and get stuck to the electrodes, thus preventing the harmful particles from exiting with the exhaust.

Here, you will write a function to compute the vertical position of a particle as it passes through the precipitator and then design some test cases to verify that your function works correctly.

![Fig 2](images/esp.png "Title")

Fig 2. Example of charged particle movement between high voltage electrodes. Particles enter through the opening on the left and are attracted to the oppositely charged electrode. In this example, a positively charged particle enters in the middle of the two electrodes and is accelerated towards the negatively charged electrode by the coulomb force, Fc. When the particle reaches the electrode, it sticks to the surface and stop moving.

If we assume this force is constant, the vertical position of the particle can be approximated using the following equation:
$$
p(t) = -\frac{1}{20000} \cdot \frac{qE}{m} t^2 + p_{\text{init}}
$$


| Variable     | Description                                           | Units                         |
|--------------|-------------------------------------------------------|-------------------------------|
| $q$          | Particle charge                                       | Nanocoulombs (nC)             |
| $E$          | Electric field strength between the electrodes        | Kilonewtons per coulomb (kN/C)|
| $m$          | Mass of the particle                                  | Nanograms (ng)                |
| $t$          | Time since the particle entered the air channel       | Microseconds (µs)             |
| $p_{\text{init}}$ | Initial vertical position of the particle      | Centimetres (cm)              |
| $p(t)$       | Vertical position of the particle at time `t`       | Centimetres (cm)              |

---


## Programming task
In the next cell, complete the function `compute_particle_height`. The function should calculate and *return* the vertical position of a particle within the electrostatic precipitator. The function accepts five input parameters:

- `q` – The charge of the particle in nanocoulombs
- `E` – The electric field strength in kilonewtons per coulomb
- `m` – The mass of the particle in nanograms
- `L` – The distance between the two electrodes in centimetres
- `t` – The time since the particle entered the air channel in microseconds

### Assumptions:
- The botton electrode is at position 0 cm and the top electrode is at position `L` cm. `L` will always be greater than zero.
- The bottom electrode will be negatively charged and the top will be positvely charged.
- Particles enter the channel halfway between the two plates (i.e., $p_{\text{init}} = L/2$ cm).
- Particles have zero vertical velocity when entering the air channel.
- __Particles immediately stop moving when contacting one of the electrodes.__
- All other forces (gravity, drag, etc.) are negligible.

Your function should round the result to **3 decimal places**.

**Hint**: This problem is designed to be solved without the use of conditional (if) statements (we'll get to those in a couple of weeks). Instead, for this lab, you may find the built-in Python functions `max` and `min` helpful. Try running
`help(min)` and 
`help(max)`
in the cell below to find out more about these functions.


In [12]:
def compute_particle_height(q,E,m,t,L):
    """
    (float, float, float, float, float) -> float
    
    Calculate the vertical height of a charged particle within an electrostatic precipitator.

    Parameters
    ----------
    q : float
        Charge of the particle in nanocoulombs (nC).
    E : float
        Electric field strength in kilonewtons per coulomb (kN/C).
    m : float
        Mass of the particle in nanograms (ng).
    t : float
        Time since the particle entered the precipitator in microseconds (µs).
    L : float
        Distance between the parallel plate electrodes in centimetres (cm).

    Returns
    -------
    float
        The height of the particle in centimetres (cm).

    Examples
    --------
    >>> compute_particle_height(0, 150, 9.2, 3.6, 5.0)
    2.5

    >>> compute_particle_height(2.3, 150, 9.2, 26.8, 5.0)
    1.153

    >>> compute_particle_height(-2.3, 160, 9.2, 36.8, 5.0)
    5.0
    """
    initial_vertical_position = L / 2
    vertical_position = (-1 / 20000) * ((q * E) / m) * t ** 2 + initial_vertical_position
    vertical_position_3_decimal = round(vertical_position, 3)
    return max(min(vertical_position_3_decimal, L), 0.0)

    
# help(max)
# help(min)


## Testing your function
As before, you will need to test your function to verify its correctness. Let's start with a simple test.

In [13]:
## Run a simple test
# step 1: execute the function and assign the return value to a variable
particle_position = compute_particle_height(2.3, 150, 9.2, 26.8, 5.0)
# step 2: display the result
print("The function returned: ", particle_position)

The function returned:  1.153


#### It runs! =D ... or maybe it doesn't :/
If your code crashed (you got an error message), you will need to correct the errors and then try running the test again.

Once your function runs without error and returns the correct value, we can move on to designing more test cases.

In [14]:
## Test #2 - Larger particle
# Double the mass of the particle
# step 1: execute the function and assign the return value to a variable
particle_position = compute_particle_height(2.3, 150, 18.4, 26.8, 5.0)
# step 2: display the result
print("The function returned: ", particle_position)

The function returned:  1.827


Hopefully, your function returns `1.827`. You can verify that is the correct value by plugging the values into the equation above. If your function is returning the correct value, great! But should we feel more confident that our function is correct? Unfortunately, not really. Even though we changed the input and got a different result, all we're testing is whether our function implemented the equation correctly. We have not tested the other components of our function. Let's look at another example to understand what we mean. This time, let's increase the time parameter significantly to `260.8`.

In [15]:
## Test #2 - Larger particle
# Increase the time
# step 1: execute the function and assign the return value to a variable
particle_position = compute_particle_height(2.3, 150, 18.4, 260.8, 5.0)
# step 2: display the result
print("The function returned: ", particle_position)

The function returned:  0.0


If we plug this into our equation we get:
$$
p(t) = -\frac{1}{20000} \cdot \frac{qE}{m} t^2 + p_{\text{init}} = -\frac{1}{20000} \cdot \frac{2.3\cdot150}{18.4}\cdot260.8^2 + 2.5 = -61.266
$$

But, one of the assumptions for our function was: "__Particles immediately stop moving when contacting one of the electrodes__". This means that our positive particle from this test stops moving when it hits the bottom electrode and the height cannot be lower than zero. So, the value returned by our function should be `0.0` and not `-61.266`.

This test case represents a different scenario because, unlike before, the output is not simply the result of plugging values into our equation. Instead, it repesents a scenario where the particle hits one of the boundaries. Identifying these different scenarios is the key to effectively testing your code.

#### Other scenarios
Some other scenarios that we could consider for this function include:
 - The charge of the particle (does the particle move towards the top boundary when a negative charge is passed to the function? what if the particle has zero charge?)
 - Does the function return the correct height when the particle makes contact with the top electrode?
 
 
Now design some of your own test cases, and try them out in the cell below.

In [16]:
# Write your own test cases here
test1 = compute_particle_height(-8.7, 160, 9.2, 36.8, 7.0) # should output 7.0
print(test1)

test2 = compute_particle_height(8.7, 160, 9.2, 36.8, 7.0) # should output 0.0
print(test2)

test3 = compute_particle_height(0.7, 80, 3.2, 36.8, 3.0) # should output 0.315
print(test3)

test4 = compute_particle_height(0.0, 80, 3.2, 36.8, 6.0) # should output 3.0
print(test4)

7.0
0.0
0.315
3.0


Once you have thoroughly tested your code, submit this notebook to Gradescope.